apps/insights/src/context/pyth-pro-demo/pyth-pro-app-state.tsx

```diff
@@ -111,21 +167,56 @@ export function PythProAppStateProvider({ children }: PropsWithChildren) {
       out = Object.values(DATA_SOURCES_EQUITY.Values);
     } else if (isAllowedFutureSymbol(appState.selectedSource)) {
       out = Object.values(DATA_SOURCES_FUTURES.Values);
+    } else if (isReplaySymbol(appState.selectedSource)) {
+      out = Object.values(DATA_SOURCES_HISTORICAL.Values);
     }
     return out.sort();
   }, [appState.selectedSource]);
 
+  const handleToggleDataSourceVisibility = useCallback<
+    AppStateContextVal["handleToggleDataSourceVisibility"]
+  >((datasource) => {
+    setDataSourceVisibility((prev) => ({
+      ...prev,
+      [datasource]: !prev[datasource],
+    }));
+  }, []);
+
+  const handleSelectPlaybackSpeed = useCallback<
+    AppStateContextVal["handleSelectPlaybackSpeed"]
+  >((speed) => {
+    setPlaybackSpeed(speed);
+  }, []);
+
   /** provider val */
   const providerVal = useMemo<AppStateContextVal>(
     () => ({
       ...appState,
       addDataPoint,
       dataSourcesInUse,
+      dataSourceVisibility,
+      handleSelectPlaybackSpeed,
       handleSelectSource,
+      handleToggleDataSourceVisibility,
+      playbackSpeed,
     }),
-    [addDataPoint, appState, dataSourcesInUse, handleSelectSource],
+    [
+      appState,
+      addDataPoint,
+      dataSourcesInUse,
+      dataSourceVisibility,
+      handleSelectPlaybackSpeed,
+      handleSelectSource,
+      handleToggleDataSourceVisibility,
+      playbackSpeed,
+    ],
   );
 
+  /** effects */
+  useEffect(() => {
+    selectedSymbolRef.current = appState.selectedSource;
```

Comment:
I'm not sure I fully understand what's happening with this ref, can you maybe elaborate?  If it's doing what I think it's doing, can you use something like [usePrevious](https://react-hookz.github.io/web/?path=/docs/state-useprevious--example) to make it more obvious?